elks/tools/mtools/mtools-4.0.23/mclasserase.1
'\" t
.TH mclasserase 1 "10Dec18" mtools-4.0.23
.SH Name
mclasserase - erase memory cards
'\" t
.de TQ
.br
.ns
.TP \\$1
..

.tr \(is'
.tr \(if`
.tr \(pd"

.SH Note\ of\ warning
This manpage has been automatically generated from mtools's texinfo
documentation, and may not be entirely accurate or complete.  See the
end of this man page for details.
.PP
.SH Description
.PP
The \fR\&\f(CWmclasserase\fR command is used to wipe memory cards by
overwriting it three times: first with \fR\&\f(CW0xff\fR, then with
\&\fR\&\f(CW0x00\fR, then with \fR\&\f(CW0xff\fR again. The command uses the following
syntax:
.PP
 
.nf
.ft 3
.in +0.3i
\&\fR\&\f(CWmclasserase [\fR\&\f(CW-d] \fImsdosdrive\fR\&\f(CW
.fi
.in -0.3i
.ft R
.PP
 
\&\fR
.PP
MS-DOS drive is optional, if none is specified, use \fR\&\f(CWA:\fR. If more than
one drive are specified, all but the last are ignored.
.PP
\&\fR\&\f(CWMclasserase\fR accepts the following command line options:
.TP
\&\fR\&\f(CWd\fR\ 
Stop after each erase cycle, for testing purposes
.TP
\&\fR\&\f(CWp\fR\ 
Not yet implemented
.PP
\&\fR\&\f(CWMclasserase\fR returns 0 on success or -1 on failure.
.PP
.SH See\ Also
Mtools' texinfo doc
.SH Viewing\ the\ texi\ doc
This manpage has been automatically generated from mtools's texinfo
documentation. However, this process is only approximative, and some
items, such as crossreferences, footnotes and indices are lost in this
translation process.  Indeed, these items have no appropriate
representation in the manpage format.  Moreover, not all information has
been translated into the manpage version.  Thus I strongly advise you to
use the original texinfo doc.  See the end of this manpage for
instructions how to view the texinfo doc.
.TP
* \ \ 
To generate a printable copy from the texinfo doc, run the following
commands:
 
.nf
.ft 3
.in +0.3i
    ./configure; make dvi; dvips mtools.dvi
.fi
.in -0.3i
.ft R
.PP
 
\&\fR
.TP
* \ \ 
To generate a html copy,  run:
 
.nf
.ft 3
.in +0.3i
    ./configure; make html
.fi
.in -0.3i
.ft R
.PP
 
\&\fRA premade html can be found at
\&\fR\&\f(CW\(ifhttp://www.gnu.org/software/mtools/manual/mtools.html\(is\fR
.TP
* \ \ 
To generate an info copy (browsable using emacs' info mode), run:
 
.nf
.ft 3
.in +0.3i
    ./configure; make info
.fi
.in -0.3i
.ft R
.PP
 
\&\fR
.PP
The texinfo doc looks most pretty when printed or as html.  Indeed, in
the info version certain examples are difficult to read due to the
quoting conventions used in info.
.PP


elks/CHANGELOG
This file is used to record changes made to the CVS database. Please add to
the top of this file before committing any changes. Please use the header
format provided for each entry, noting that the text on each line should not
be longer than the lines of dashes.

Al Riddoch.

Contributor           Email address                     CHANGELOG entries found
~~~~~~~~~~~~~~~~~~~~  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~  ~~~~~~~~~~~~~~~~~~~~~~~
Alan Cox              alan@redhat.com                   17 Jun 2001-18 Jun 2001
Alastair Bridgewater  nyef@softhome.net                 19 Nov 1998- 6 Dec 1998
Alistair Riddoch      alriddoch@zepler.org              25 Sep 1998-17 Jun 2001
Arnaud Launay         asl@launay.org                    15 Jul 1999-17 Oct 2001
Christian Mardmuller  chm@kdt.de                        10 Nov 1999-29 Feb 2000
Christoph Plattner    Christoph.Plattner@dot.at         31 Jan 2000- 2 Feb 2000
Claudio Matsuoka      claudio@conectiva.com
Daniele Giacomini     daniele@swlibero.org		20 Sep 2004
David Murn            scuffer@hups.apana.org.au         11 Nov 1998-15 Nov 1999
Eduardo Habkost	      ehabkost@conectiva.com.br		01 Jun 2004
Giles Russell                                           22 Nov 1999
Greg Haerr                                              14 Jul 1999-20 Jul 1999
Harry Kalogirou       HarKal@gmx.net                    20 Jun 2001- 2 Jan 2003
Jacek Lipkowski       sq5bpf@acid.ch.pw.edu.pl          20 Jan 2002
Jay Barr                                                27 Oct 2002
Ken Martwick          kenm@efn.org                      12 May 2002
M Warner Losh         imp@village.org                   20 Jan 2002
Manuel Novoa III      mjn3@codepoet.org                 18 Jun 2002-25 Jun 2002
Mario Frasca                                            20 Jan 2000-14 Feb 2000
Miguel Bolanos	      mike@linuxlabs.com		20 Sep 2004
Paul Nasrat           pauln@truemesh.com                29 Oct 2001-12 Jul 2002
Per Olofsson          MagerValp@cling.gu.se             25 Feb 2000
Peter L Vachuska      pvachusk@nconnect.net             13 Aug 2001
Richard Wallman       murkygoth@murkygoth.uklinux.net    2 Apr 2002
Riley Williams        Riley@Williams.Name               16 Nov 1998-13 Jul 2003
Roman Festchook       roma@polesye.net                  27 Oct 2001
Simon Weijgers        simon@weijgers.com                16 Sep 1998-15 Nov 1999
Simon Wood                                               6 Jan 2000-12 Jan 2000
Thomas McWilliams     tgm9999@yahoo.com                 27 Aug 1999- 2 Apr 2001
Tommy McCabe 	      rocketjet314@yahoo.com		31 May 2004

-------------------------------------------------------------------------------

Thu Feb 16 17:03:00 EST 2012	Jody Bruchon <jody@jodybruchon.com>	
	* Removed elksfs. It was a copy of the MINIX fs and no tools
	  exist to make such a filesystem.

Thu Nov 23 21:35:28 EST 2006	Segin <segin2005@gmail.com>
	* Changed name for the submenu for selecting charater
	  device drivers to note the fact that that submenu is broken;
	  attempting to enter it will have lxdialog die horribly.

Wed Nov  8 17:49:52 EST 2006	Segin <segin2005@gmail.com>

	* Changed ELKS init sequence to check all locations for 'init'
	  that the Linux kernel does. This is to follow more 
	  "traditional" behaviour.

Sun Aug 20 21:51:25 UTC 2006	Claudio Matsuoka <claudio@mandriva.com>

	* Allow loading Minix executables with total size larger than
	  64KB, such as ash (70140 bytes). Currently the total size
	  minus 32 bytes of header was limited to 16-bit.

	* Set tty1 back to minor 0. This is different from Linux
	  numbering, but it's necessary to have current ELKS virtual
	  terminals working. This and the executable size change put
	  ELKS back in usable state, allowing logins in the first VT
	  and executing a shell.

Mon Sep 20 23:27:23 CST 2004	Daniele Giacomini <daniele@swlibero.org>

	* Added italian keys support, Miguel Bolanos cleaned the original
	  file code.

Wed Jun  9 17:27:52 BRT 2004	Eduardo Habkost <ehabkost@conectiva.com.br>

	* fix for my last fix: s/rp_cmd/rq_cmd/

Thu Jun 03 21:47:13 CST 2004    Claudio Matsuoka <claudio@conectiva.com>

	* fixed format specifiers to correctly displaysegment sizes as unsigned
	  values, and to display debug messages correctly.

Thu Jun 03 07:48:25 CST 2004	Eduardo Habkost <ehabkost@conectiva.com.br>

	* Fixed wrong struct field name for arch/i86/drivers/block/doshd.c
	* Several fixes ro arch/i86/mm/user.c

Web Jun 02 14:16:57 CST 2004    Claudio Matsuoka <claudio@conectiva.com>

	* Added an extra fix to Eduardos memcpy fix.

Tue Jun 01 21:00:03 CST 2004	Eduardo Habkost ehabkost@conectiva.com.br

	* Fix memcpy_{to,from}fs, adding code to save the contents of SI and DI

Mon May 31 07:39:47 CST 2004	Tommy McCabe <rocketjet314@yahoo.com>

	* Fix wrong assignment (typo?) on printk
	* Fix order of LD parameters on arch/i86/Makefile
	* Use #ifdef rightly on setup.S
	* Fix init to use tty0, not tty1 as system console.

===============================================================================

Sun Jul 13 04:24:44 BST 2003	Riley Williams <Riley@Williams.Name>

	* This is ELKS 0.1.3-pre1

Sat Jul 12 12:24:36 BST 2003	Riley Williams <Riley@Williams.Name>

	* Rewrote kernel/printk.c to allow printk() to correctly print out
	  signed values.
	* Rewrote kernel/printk.c to produce a less verbose panic() dump,
	  and also to increase the number of lines of stack trace to 10.
	* Various spelling corrections to printk() messages.

===============================================================================

Thu Jan 02 16:12:36 EET 2003	Harry Kalogirou <harkal@gmx.net>

        * Better select hash.
	* This is ELKS 0.1.2

-------------------------------------------------------------------------------

Fri Nov 01 01:52:44 EET 2002	Harry Kalogirou <harkal@gmx.net>

        * Changed the position of ds in the _registers struct.
        * Added support for loading the new ELKS executable format. This is 

elks/tools/mtools/mtools-4.0.23/mclasserase.c
/*  Copyright 2003 Stefan Feuz, Lukas Meyer, Thomas Locher
 *  Copyright 2004,2006,2007,2009 Alain Knaff.
 *  This file is part of mtools.
 *
 *  Mtools is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  Mtools is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with Mtools.  If not, see <http://www.gnu.org/licenses/>.
 *
 * Filename:
 *    mclasserase.c
 *
 * Original Creation Date:
 *    05.III.2003
 *
 * Copyright:
 *    GPL
 *
 * Programmer:
 *    Stefan Feuz, Lukas Meyer, Thomas Locher
 */

#include "sysincludes.h"
#include "msdos.h"
#include "mtools.h"
#include "vfat.h"
#include "mainloop.h"
#include "fsP.h"

#ifdef HAVE_GETOPT_H
#include <getopt.h>
#endif

#include "file.h"

#include <unistd.h>
#include <stdio.h>

/**
 * Prints the Usage Message to STDOUT<br>
 *
 * @author  stefan feuz<br>
 *          stefan.feuz@ruag.com
 *
 * @param   n.a.
 *
 * @returns n.a.
 *
 */
static void usage(int ret) NORETURN;
static void usage(int ret)
{
  fprintf(stderr, "Mtools version %s, dated %s\n", mversion, mdate);
  fprintf(stderr, "Usage: %s [-d] drive:\n", progname);
  exit(ret);
}

/**
 * Delete all files on a Drive.<br>
 *
 * @author  Lukas Meyer<br>
 *          lukas.meyer@ruag.com
 * @version 0.4, 11.12.2003
 *
 * @param   drive   the drive to erase
 * @param   debug   1: stop after each erase cycle, 0: normal mode
 * 
 * @returns n.a.
 *
 */
static void do_mclasserase(char drive,int debug) NORETURN;
static void do_mclasserase(char drive,int debug)
{
  struct device dev;		/* Device information structure */
  union bootsector boot;

  int media;			/* Just used to enter some in find_device */
  char name[EXPAND_BUF];
  Stream_t *Stream;
  struct label_blk_t *labelBlock;
  
  FILE * fDevice;              /* Stores device's file descriptor */
  
  char cCardType[12];
  
  char drivel[3];		/* Stores the drive letter */


  int i = 0;

  /* FILE *forf; */

  char dummy[2];       /* dummy input for debugging purposes.. */
  int icount=0;
  int iTotalErase = 0;

/* How many times we'll overwrite the media: */
#define CYCLES 3
  unsigned char odat[CYCLES];	/* Data for each overwrite procedure */

  /* Creating values for overwrite  */
  odat[0]=0xff;
  odat[1]=0x00;
  odat[2]=0xff;
  

  if (debug == 1)
     printf("cycles: %i, odats: %i,%i,%i\n",CYCLES,odat[0],odat[1],odat[2]);
  
  

  /* Reading parameters from card. Exit with -1 if failed. */
  if(! (Stream = find_device(drive, O_RDONLY, &dev, &boot,
					   name, &media, 0, NULL)))
	exit(1);

  FREE(&Stream);

  /* Determine the FAT - type */
#if 0
  if(WORD(fatlen)) {
    labelBlock = &bbelBlock = &boot->ext.old.labelBlock;
  } else {
    labelBlock = &boot->ext.fat32.labelBlock;
  }
#endif

  /* we use only FAT12/16 ...*/
  labelBlock = &boot.boot.ext.old.labelBlock;
   
  /* store card type */
  sprintf(cCardType, "%11.11s", labelBlock->label);

  if (debug == 1)
  {
     printf("Using Device: %s\n",name);
     printf("Card-Type detected: %s\n",cCardType);
  }
      
  /* Forming cat command to overwrite the medias content. */
  sprintf( drivel, "%c:", ch_tolower(drive) );

#if 0
  media_sectors = dev.tracks * dev.sectors;
  sector_size = WORD(secsiz) * dev.heads; 
  

	printf(mcat);
	printf("\n%d\n", media_sectors);
	printf("%d\n", sector_size);
#endif
 
  /*
   * Overwrite device
   */
  for( i=0; i < CYCLES; i++){

     if (debug==1)
     {
        printf("Erase Cycle %i, writing data: 0x%2.2x...\n",i+1,odat[i]);
     }

     fDevice = fopen(name,"ab+");
          
     if (fDevice == 0)
     {
        perror("Error opening device");
	exit(-1);
     }


     if (debug==1)
     {
        printf("Open successful...\n");
	printf("Flushing device after 32 kBytes of data...\n");
	printf("Erasing:");
	fflush( stdout );
     }

     /* iTotalErase = 0; */
      
     /*
      * overwrite the whole device
      */
     while ((feof(fDevice)==0) && (ferror(fDevice)==0))
     {
          		
	fputc(odat[i],fDevice);

	icount++;
	if (icount > (32 * 1024))
	{
	   /* flush device every 32KB of data...*/
	   fflush( fDevice );
           
	   iTotalErase += icount;
	   if (debug == 1)
	   {
	      printf(".");	   
	      fflush( stdout );
	   }
	   icount=0;
	}
     }
     
     if (debug==1)
     {
        printf("\nPress <ENTER> to continue\n");
        printf("Press <x> and <ENTER> to abort\n");
	
	if(scanf("%c",dummy) < 1)
	  printf("Input error\n");
	fflush( stdin );
	
	if (strcmp(dummy,"x") == 0)
	{
	   printf("exiting.\n");
	   exit(0);
	}
     }
     
     fclose(fDevice);

  }

   
  /*
   * Format device using shell script
   */  
  if (debug == 0)
  {
  	/* redirect STDERR and STDOUT to the black hole... */
	if (dup2(open("/dev/null", O_WRONLY), STDERR_FILENO) != STDERR_FILENO)
		printf("Error with dup2() stdout\n");
	if (dup2(open("/dev/null", O_WRONLY), STDOUT_FILENO) != STDOUT_FILENO)
	        printf("Error with dup2() stdout\n");
  }
	       
  if (debug == 1)
      printf("Calling amuFormat.sh with args: %s,%s\n",cCardType,drivel);
	
  execlp("amuFormat.sh","",cCardType,drivel,NULL);

  /* we never come back...(we shouldn't come back ...) */
  exit(-1);
    
}


/**
 * Total Erase of Data on a Disk. After using mclasserase there won't
 * be ANY bits of old files on the disk.<br>
 * </b>
 * @author  stefan feuz<br>
 *          thomas locher<br>
 *          stefan.feuz@ruag.com
 *          thomas.locher@ruag.com
 * @version 0.3, 02.12.2003
 *
 * @param   argc    generated automatically by operating systems
 * @param   **argv1  generated automatically by operating systems
 * @param   type    generated automatically by operating systems
 *
 * @param   -d      stop after each erase cycle, for testing purposes
 * 
 * @returns int     0 if all is well done<br>
 *          int     -1 if there is something wrong
 *
 * @info    mclasserase [-p tempFilePath] [-d] drive:
 *
 *
 */
void mclasserase(int argc, char **argv, int type UNUSEDP) NORETURN;
void mclasserase(int argc, char **argv, int type UNUSEDP)
{
  /* declaration of all variables */
  int c;
  int debug=0;
  /* char* tempFilePath=NULL; */
  char drive='a';

  extern int optind;

  destroy_privs();

  /* check and read command line arguments */
#ifdef DEBUG
  printf("mclasserase: argc = %i\n",argc);
#endif
  /* check num of arguments */
  if(helpFlag(argc, argv))
    usage(0);
  if ( (argc != 2) & (argc != 3) & (argc != 4))
    { /* wrong num of arguments */
    printf ("mclasserase: wrong num of args\n");
    usage(1);
  }
  else
  { /* correct num of arguments */
    while ((c = getopt(argc, argv, "+p:dh")) != EOF)
    {
      switch (c)
      {
	
	case 'd':
           
	   printf("=============\n");
	   printf("Debug Mode...\n");
	   printf("=============\n");
           debug = 1;
	   break;
	case 'p':
           printf("option -p not implemented yet\n"); 
           break;
	case 'h':
	  usage(0);
        case '?':
           usage(1);
        default:
           break;
       }
     }
#ifdef DEBUG
     printf("mclasserase: optind = %i\n",optind);
   /* look for the drive to erase */
   printf("mclasserase: searching drive\n");
#endif
   for(; optind < argc; optind++)
   {
     if(!argv[optind][0] || argv[optind][1] != ':')
     {
       usage(1);
     }
     drive = ch_toupper(argv[optind][0]);
   }
  }
#ifdef DEBUG
  printf("mclasserase: found drive %c\n", drive);
#endif 
  /* remove all data on drive, you never come back if drive does
   * not exist */
  
  do_mclasserase(drive,debug);
}
